lib/category_badge.rb

```diff
@@ -4,48 +4,93 @@ module CategoryBadge
   def self.html_for(category, opts = nil)
     opts = opts || {}
 
-    # If there is no category, bail
-    return "" if category.blank?
+    # Bail if there is no category, hide uncategorized by default
+    return "" if category.blank? || (category.uncategorized? && !opts[:show_uncategorized])
 
-    # By default hide uncategorized
-    return "" if category.uncategorized? && !opts[:show_uncategorized]
+    if opts[:inline_style]
+      # Inline styles for email
+      style_for_email(category, opts)
+    else
+      # Browser styles
+      style_for_browser(category, opts)
+    end
+  end
 
-    extra_classes = "#{opts[:extra_classes]}"
+  def self.shared_data(category, opts)
+    {
+      parent_category: fetch_parent_category(category),
+      category_url:
+        opts[:absolute_url] ? "#{Discourse.base_url_no_prefix}#{category.url}" : category.url,
+      extra_classes: opts[:extra_classes].to_s,
+    }
+  end
 
-    result = +""
+  def self.fetch_parent_category(category)
+    Category.find_by(id: category.parent_category_id) unless category.parent_category_id.nil?
```

Comment:
```suggestion
    Category.find_by(id: category.parent_category_id) if category.parent_category_id
```

Not new to your code but easier to read.